base/contextvariables.jl

```diff
@@ -0,0 +1,415 @@
+    )
+end
+
+"""
+    with_context(f, var1 => value1, var2 => value2, ...)
+    with_context(f, pairs)
+
+Run `f` in a context with given values set to the context variables.  Variables
+specified in this form are rolled back to the original value when `with_context`
+returns.  It act like a dynamically scoped `let`.  If `nothing` is passed as
+a value, corresponding context variable is cleared; i.e., it is unassigned or
+takes the default value.  Use `Some(value)` to set `value` if `value` can be
+`nothing`.
+
+    with_context(f, nothing)
+
+Run `f` in a new empty context.  All variables are rewind to the original values
+when `with_context` returns.
+
+Note that
+
+```julia
+var2[] = value2
+with_context(var1 => value1) do
+    @show var2[]  # shows value2
+    var3[] = value3
+end
+@show var3[]  # shows value3
+```
+
+and
+
+```julia
+var2[] = value2
+with_context(nothing) do
+    var1[] = value1
+    @show var2[]  # shows default (or throws)
+    var3[] = value3
+end
+@show var3[]  # does not show value3
+```
+
+are not equivalent.
+"""
+function with_context(f, kvs::Pair{<:ContextVar}...)
+    orig = map(((k, _),) -> k => get(k), kvs)
+    set_context(kvs)
+    try
+        return f()
+    finally
+        set_context(orig)
+    end
+end
+
+function with_context(f, ::Nothing)
+    ctx = get_task_ctxvars()
+    try
+        set_task_ctxvars(nothing)
+        return f()
+    finally
+        set_task_ctxvars(ctx)
+    end
+end
+
+# Not using `set_context!` so that `snapshot` as an input makes sense.
+"""
+    set_context(var1 => value1, var2 => value2, ...)
+    set_context(pairs)
+    set_context(snapshot::ContextSnapshot)
+
+Equivalent to `var1[] = value1`, `var2[] = value2`, and so on.  The second form
+expect an iterable of pairs of context variable and values.  This function is
+more efficient than setting individual context variables sequentially.  Like
+[`with_context`](@ref), `nothing` value means to clear the context variable
+and `Some` is always unwrapped.
+
+A "snapshot" of the context returned from [`snapshot_context`](@ref) can
+be specified as an input (the third form).
+
+# Examples
+```jldoctest
+julia> @contextvar global x
+       @contextvar global y;
+
+julia> set_context(x => 1, y => 2)
+
+julia> x[]
+1
+
+julia> y[]
+2
+
+julia> set_context([x => 10, y => 20])
+
+julia> x[]
+10
+
+julia> y[]
+20
+```
+"""
+set_context(kvs::Pair{<:ContextVar}...) = set_context(kvs)
+function set_context(kvs)
+    ctx = merge_ctxvars(get_task_ctxvars(), (k.key => v for (k, v) in kvs))
+    set_task_ctxvars(ctx)
+    return
+end
+
+struct ContextSnapshot{T}
+    vars::T
+end
+
+# TODO: Do we need to implement `Dict{ContextVar}(::ContextSnapshot)`?
+# This requires storing UUID-to-ContextVar mapping somewhere.
+
+"""
+    snapshot_context() -> snapshot::ContextSnapshot
+
+Get a snapshot of a context that can be passed to [`reset_context`](@ref) to
+rewind all changes in the context variables.
```

Comment:
It is called [`contextvars.copy_context`](https://docs.python.org/3/library/contextvars.html#contextvars.copy_context) in Python and [`ExecutionContext.CreateCopy`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.executioncontext.createcopy?view=netcore-3.1) in .NET.  But I was not sure if "copy" is the right word when the underlying data is (treated as) immutable.